test/e2e/storage/testsuites/volume_io.go

```diff
@@ -0,0 +1,364 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+ * This test checks that the plugin VolumeSources are working when pseudo-streaming
+ * various write sizes to mounted files.
+ */
+
+package testsuites
+
+import (
+	"fmt"
+	"math"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+// MD5 hashes of the test file corresponding to each file size.
+// Test files are generated in testVolumeIO()
+// If test file generation algorithm changes, these must be recomputed.
+var md5hashes = map[int64]string{
+	testpatterns.FileSizeSmall:  "5c34c2813223a7ca05a3c2f38c0d1710",
+	testpatterns.FileSizeMedium: "f2fa202b1ffeedda5f3a58bd1ae81104",
+	testpatterns.FileSizeLarge:  "8d763edc71bd16217664793b5a15e403",
+}
+
+type volumeIOTestSuite struct {
+	name         string
+	testPatterns []testpatterns.TestPattern
+}
+
+var _ TestSuite = &volumeIOTestSuite{}
+
+// InitVolumeIOTestSuite returns volumeIOTestSuite that implements TestSuite interface
+func InitVolumeIOTestSuite() TestSuite {
+	return &volumeIOTestSuite{
+		name: "volumeIO",
+		testPatterns: []testpatterns.TestPattern{
+			testpatterns.DefaultFsInlineVolume,
+			testpatterns.DefaultFsPreprovisionedPV,
+			testpatterns.DefaultFsDynamicPV,
+		},
+	}
+}
+
+func (t *volumeIOTestSuite) getName() string {
+	return t.name
+}
+
+func (t *volumeIOTestSuite) getTestPatterns() []testpatterns.TestPattern {
+	return t.testPatterns
+}
+
+func (t *volumeIOTestSuite) skipUnsupportedTest(tp testpatterns.TestPattern, drv drivers.TestDriver) {
+}
+
+func (t *volumeIOTestSuite) initTestResource() TestResource {
```

Comment:
I'm not convinced that we need to separate out resource creation from test input creation.  Can createVolumeIOTestInput just call the genericVolumeTestResource.setupResource()?

Reply:
Fixed as suggested. PTAL